cardano-api/src/Cardano/Api/DeserialiseAnyOf.hs

```diff
@@ -0,0 +1,272 @@
+          DeserialiseInputError (InputTextEnvelopeError err)
+
+        -- The input was not valid a text envelope.
+        Left _ -> DeserialiseInputErrorFormatMismatch
+
+    deserialiseBech32 :: SerialiseAsBech32 a => DeserialiseInputResult a
+    deserialiseBech32 =
+      case deserialiseFromBech32 asType inputText of
+        Right res -> DeserialiseInputSuccess res
+
+        -- The input was not valid Bech32.
+        Left (Bech32DecodingError _) -> DeserialiseInputErrorFormatMismatch
+
+        -- The input was valid Bech32, but some other error occurred.
+        Left err -> DeserialiseInputError $ InputBech32DecodeError err
+
+    deserialiseHex :: SerialiseAsRawBytes a => DeserialiseInputResult a
+    deserialiseHex
+      | isValidHex inputBs =
+          maybe
+            (DeserialiseInputError InputInvalidError)
+            DeserialiseInputSuccess
+            (deserialiseFromRawBytesHex asType inputBs)
+      | otherwise = DeserialiseInputErrorFormatMismatch
+
+    isValidHex :: ByteString -> Bool
+    isValidHex x =
+      all (`elem` hexAlpha) (toLower <$> BSC.unpack x)
+        && even (BSC.length x)
+      where
+        hexAlpha :: [Char]
+        hexAlpha = "0123456789abcdef"
+
+-- | Deserialise an input of some type that is formatted in some way.
+--
+-- The provided 'ByteString' can either be Bech32-encoded or in the text
+-- envelope format.
+deserialiseInputAnyOf
+  :: forall b.
+     [FromSomeType SerialiseAsBech32 b]
+  -> [FromSomeType HasTextEnvelope b]
+  -> ByteString
+  -> Either InputDecodeError b
+deserialiseInputAnyOf bech32Types textEnvTypes inputBs =
+    case deserialiseBech32 `orTry` deserialiseTextEnvelope of
+      DeserialiseInputSuccess res -> Right res
+      DeserialiseInputError err -> Left err
+      DeserialiseInputErrorFormatMismatch -> Left InputInvalidError
+  where
+    inputText :: Text
+    inputText = Text.decodeUtf8 inputBs
+
+    orTry
+      :: DeserialiseInputResult b
+      -> DeserialiseInputResult b
+      -> DeserialiseInputResult b
+    orTry x y =
+      case x of
+        DeserialiseInputSuccess _ -> x
+        DeserialiseInputError _ -> x
+        DeserialiseInputErrorFormatMismatch -> y
+
+    deserialiseTextEnvelope :: DeserialiseInputResult b
+    deserialiseTextEnvelope = do
+      let textEnvRes :: Either TextEnvelopeError b
+          textEnvRes =
+            deserialiseFromTextEnvelopeAnyOf textEnvTypes
+              =<< first TextEnvelopeAesonDecodeError (Aeson.eitherDecodeStrict' inputBs)
+      case textEnvRes of
+        Right res -> DeserialiseInputSuccess res
+
+        -- The input was valid a text envelope, but there was a type mismatch
+        -- error.
+        Left err@TextEnvelopeTypeError{} ->
+          DeserialiseInputError (InputTextEnvelopeError err)
+
+        -- The input was not valid a text envelope.
+        Left _ -> DeserialiseInputErrorFormatMismatch
+
+    deserialiseBech32 :: DeserialiseInputResult b
```

Comment:
I reckon make this past tense do show the deserialisation was already done.

Reply:
Yeah I agree, thanks.